### ISSUE TITLE
Singeleton pattern for the task API

### ISSUE THREAD

**Issue by sxjscience** [awslabs/autogluon#382]
Currently, in AutoGluon, we will do the following, in which `TabularPrediction` is used as a singeleton.

```python
from autogluon import TabularPrediction as task
predictor = task.fit(...)
```

A more natural design choice is to create a new class object and call fit as a method.
```python
from autogluon import TabularPrediction
task = TabularPrediction(SOME_ARGS)
predictor = task.fit()
```

@Jerryzcn

**Comment by Innixma** [comment]
I think this is reasonable, it aligns with how H2O and auto-sklearn function. It will be a major refactor in terms of user-facing API, so we should save it for a major release.

I am curious what we would gain from this outside of it being more 'natural'. I suppose we could add info/utility functions that would explain the parameters they passed in? Would .fit() still take in the data itself as a parameter?

**Comment by sxjscience** [comment]
@Innixma I think `.fit()` will still take the data as the input parameter. The difference is that we can pass some specific requirements like time budget and inference latency when constructing the `task`, e.g.

```python
from autogluon import TabularPrediction
task = TabularPrediction(time_limit=1000, latency='low')
predictor = task.fit(train_data, label='...')
```

Then, we will automatically learn a predictor that obeys the budget constraint.

**Comment by Innixma** [comment]
@jwmueller Any comments on this?

**Comment by jwmueller** [comment]
I agree the instantiating the `TabularPrediction` object is more natural Python, but what properties would be set when the object is instantiated to justify this extra line of code?  I don't think the object constructor should be tracking time-limits, those should be in `fit()` since they are never used afterward.  But the object constructor could be appropriately used for things which remain relevant to the task even after `fit()`, for example optional things like: the metric, problem_type.  Overall it would be significant work to refactor everything so IMO needs to bring some extra benefits beyond being seeming more natural (To me, requiring an extra line of code more or less cancels out the 'natural' benefit).

**Comment by sxjscience** [comment]
Yes, we can set some evaluation metrics bound to the learning task. We could also probably set 1) the IP address of the servers, 2) the directory that the task will use for storing temporary results. Putting the learning-related arguments inside `fit()` will end up with a very large fit function. Nevertheless, this is just something that would be nice to have.

**Comment by jwmueller** [comment]
In principle, I agree with including the directory since its relevant outside of `fit()`.  However, this comes with some thorny issues. Suppose the user calls `task = TabularPrediction(directory=...)` and then calls `a = task.fit(...)`, `b = task.fit(...)`.  The second `fit()` will overwrite the files of the first which may be less intuitive to the user than when they explicitly have to pass the directory to `fit()`.

I'm unsure about including the IP addresses.  Currently these servers are only used during `fit()`, and I think anything that's not used outside of `fit()` can just be kept as a `fit()` argument.  
 It's not like the servers are being created/affected by specifying their IP addresses in the proposed task constructor or any actual setup is really happening.  I'd be more interested in including IP addresses if some sort of related setup is actually happening inside the task constructor (for example, the task constructor could check that it can actually communicate with these IP addresses).

Advantages of the proposed change I see so far are:
- more natural
- can reduce number of fit() args
- can potentially do some non-fit()-related setup inside the task object construction (e.g. checking that specified IP addresses are valid)

Disadvantages:
- user has to write extra line of code to use autogluon

**Comment by sxjscience** [comment]
I find that there is actually another `Learner` class in AutoGluon, which behaves more similar to the model in scikit-learn.

**Comment by Innixma** [comment]
Learner was the original user-facing API prior to open-source, however we added Predictor as a wrapper to it to conform to the API of ImageClassification and TextClassification.

**Comment by Innixma** [comment]
Closing as resolved, we have implemented the singleton pattern for v0.1.
